src/operators.jl

```diff
@@ -0,0 +1,268 @@
+using LinearAlgebra: diagm, lmul!
+
+export raising, lowering, number, X, Y, X_Y, XY, flip_flop,
+       decay, dephasing, dipole_drive, parametric_drive, rwa_dipole
+
+######################################################
+# Primitives
+######################################################
+
+# ladder operators
+"""
+    raising(q::QSystem, ϕ::Real=0.0)
+
+raising/creation/a ladder operator. Given an eignenstate of the number operator `|n⟩` on a
+`QSystem q`, `raising(q)|n⟩ = √(n+1)|n+1⟩`. Optionally an additional phase exp(2πiϕ) is applied.
+
+# Examples
+```jldoctest
+julia> q = DuffingTransmon("q", 3, DuffingSpec(5, -0.2))
+DuffingTransmon("q", 3, DuffingSpec(5, -0.2))
+
+julia> raising(q)
+3×3 Array{Float64,2}:
+ 0.0  0.0      0.0
+ 1.0  0.0      0.0
+ 0.0  1.41421  0.0
+```
+"""
+raising(q::QSystem, ϕ::Real=0.0) = raising(dimension(q), ϕ)
+
+# custom diagm because it's twice as fast for small matrices
+function raising(dim::Integer,  ϕ::Real=0.0)
+    m = zeros(typeof(complex(ϕ)), dim, dim)
+    fac = exp(1im * 2π * ϕ)
+    for i in 1:(dim -1 )
+        m[i + 1, i] = sqrt(i) * fac
+    end
+    return m
+end
+
+"""
+    lowering(q::QSystem, ϕ::Real=0.0, factor::Real=1))
+
+lowering/destruction/a† ladder operator. Given an eignenstate of the number operator `|n⟩` on a
+`QSystem q`, `lowering(q)|n⟩ = √(n+1)|n+1⟩`. Optionally apply an additional phase ϕ and scaling `factor`.
+
+# Examples
+```jldoctest
+julia> q = DuffingTransmon("q", 3, DuffingSpec(5, -0.2))
+DuffingTransmon("q", 3, DuffingSpec(5, -0.2))
+
+julia> lowering(q)
+3×3 Array{Float64,2}:
+ 0.0  1.0  0.0
+ 0.0  0.0  1.41421
+ 0.0  0.0  0.0
+```
+"""
+lowering(q::QSystem, ϕ::Real=0, factor::Real=1) = lowering(dimension(q), ϕ, factor)
+
+# custom diagm because it's twice as fast for small matrices
+function lowering(dim::Integer,  ϕ::Real=0, factor::Real=1)
+    m = zeros(typeof(complex(float(ϕ))), dim, dim)
+    efac = exp(-1im * 2π * ϕ) * factor
+    for i in 1:(dim - 1)
+        m[i, i + 1] = sqrt(i) * efac
+    end
+    return m
+end
+
+"""
+    number(q::QSystem, factor::Real=1)
+
+Number operator (a†a) on a QSystem `q` with an optional scaling factor which gives a dephasing
+matrix if `factor != 1`.
+"""
+number(q::QSystem, factor::Real=1) = number(dimension(q), factor)
+
+# custom diagm because it's twice as fast for small matrices
+function number(dim::Integer, factor::Real=1)
+    m = zeros(factor |> float |> complex |> typeof, dim, dim)
```

Comment:
Function chaining sticks out here because it isn't used in the other similarly defined methods.

Reply:
but all the brackets we're saving...